#Yilun Wang(yilun830@bu.edu) - ASSIGNMENT #5

# convert notebook to html then print as PDF

In [2]:
!jupyter nbconvert --to html /content/Assignment_05_Yilun_Wang.ipynb


[NbConvertApp] Converting notebook /content/Assignment_05_Yilun_Wang.ipynb to html
[NbConvertApp] Writing 325194 bytes to /content/Assignment_05_Yilun_Wang.html


# Main content

## import/merge/save data

In [ ]:
%matplotlib inline

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import seaborn as sns

In [ ]:
# import temGICS.csv
temp = pd.read_csv('https://raw.githubusercontent.com/ChasteloveCNN/ba765-session02/main/tempGICSS.csv')

In [ ]:
# import project tickers.csv
data1 = pd.read_csv('https://raw.githubusercontent.com/ChasteloveCNN/ba765-session02/main/ProjectTickers.csv')

In [ ]:
temp.head()

gvkey  datadate  fyear indfmt consol popsrc datafmt  tic curcd costat  \
0   1004  20210531   2020   INDL      C      D     STD  AIR   USD      A   
1   1045  20211231   2021   INDL      C      D     STD  AAL   USD      A   
2   1075  20211231   2021   INDL      C      D     STD  PNW   USD      A   
3   1078  20211231   2021   INDL      C      D     STD  ABT   USD      A   
4   1161  20211231   2021   INDL      C      D     STD  AMD   USD      A   

   ggroup  
0    2010  
1    2030  
2    5510  
3    3510  
4    4530

In [ ]:
temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1886 entries, 0 to 1885
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   gvkey     1886 non-null   int64 
 1   datadate  1886 non-null   int64 
 2   fyear     1886 non-null   int64 
 3   indfmt    1886 non-null   object
 4   consol    1886 non-null   object
 5   popsrc    1886 non-null   object
 6   datafmt   1886 non-null   object
 7   tic       1886 non-null   object
 8   curcd     1886 non-null   object
 9   costat    1886 non-null   object
 10  ggroup    1886 non-null   int64 
dtypes: int64(4), object(7)
memory usage: 162.2+ KB


In [ ]:
temp.isna().sum()

gvkey       0
datadate    0
fyear       0
indfmt      0
consol      0
popsrc      0
datafmt     0
tic         0
curcd       0
costat      0
ggroup      0
dtype: int64

In [ ]:
data1.head()

Ticker                     Name   RetYTD
0      A     Agilent Technologies  -0.2080
1     AA               Alcoa Corp   0.4731
2    AAL     American Airlines Gp   0.0579
3    AAN  Aarons Holdings Company  -0.1327
4   AAON                 Aaon Inc  -0.3456

In [ ]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1886 entries, 0 to 1885
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Ticker  1886 non-null   object 
 1   Name    1886 non-null   object 
 2   RetYTD  1886 non-null   float64
dtypes: float64(1), object(2)
memory usage: 44.3+ KB


In [ ]:
data1.isna().sum()

Ticker    0
Name      0
RetYTD    0
dtype: int64

In [ ]:
# rename column
temp.rename(columns={'tic':'Ticker'}, inplace=True)
temp.head()

gvkey  datadate  fyear indfmt consol popsrc datafmt Ticker curcd costat  \
0   1004  20210531   2020   INDL      C      D     STD    AIR   USD      A   
1   1045  20211231   2021   INDL      C      D     STD    AAL   USD      A   
2   1075  20211231   2021   INDL      C      D     STD    PNW   USD      A   
3   1078  20211231   2021   INDL      C      D     STD    ABT   USD      A   
4   1161  20211231   2021   INDL      C      D     STD    AMD   USD      A   

   ggroup  
0    2010  
1    2030  
2    5510  
3    3510  
4    4530

In [ ]:
# merge two datasets
df = pd.merge(temp, data1, how='outer', on='Ticker')

In [ ]:
df

gvkey  datadate  fyear indfmt consol popsrc datafmt Ticker curcd  \
0       1004  20210531   2020   INDL      C      D     STD    AIR   USD   
1       1045  20211231   2021   INDL      C      D     STD    AAL   USD   
2       1075  20211231   2021   INDL      C      D     STD    PNW   USD   
3       1078  20211231   2021   INDL      C      D     STD    ABT   USD   
4       1161  20211231   2021   INDL      C      D     STD    AMD   USD   
...      ...       ...    ...    ...    ...    ...     ...    ...   ...   
1881  266315  20211231   2021   INDL      C      D     STD    KRG   USD   
1882  294524  20211231   2021   INDL      C      D     STD    LYB   USD   
1883  312009  20211231   2021   INDL      C      D     STD    FRO   USD   
1884  316056  20211231   2021   INDL      C      D     STD   ALLE   USD   
1885  317264  20210331   2020   INDL      C      D     STD    LPG   USD   

     costat  ggroup                          Name   RetYTD  
0         A    2010                      AAR Corp   0.2944  
1         A    2030          American Airlines Gp   0.0579  
2         A    5510    Pinnacle West Capital Corp   0.0985  
3         A    3510           Abbott Laboratories  -0.1638  
4         A    4530             Adv Micro Devices  -0.3533  
...     ...     ...                            ...     ...  
1881      A    6010       Kite Realty Group Trust   0.0275  
1882      A    1510  Lyondellbasell Industries NV   0.1664  
1883      A    1010                 Frontline Ltd   0.3380  
1884      A    2010                  Allegion Plc  -0.1888  
1885      A    1010                Dorian Lpg Ltd   0.2427  

[1886 rows x 13 columns]

In [ ]:
# save dataframe to csv
df.to_csv("assign5.csv")

## Calculations

### calculate average stock returns for each industry classification (24 industry groups)

In [ ]:
df['ggroup']

In [ ]:
avg = df['RetYTD'].groupby(df['ggroup']).mean()

In [ ]:
# this is the group of each industry's average stock return
avg

ggroup
1010    0.515035
1510    0.057846
2010   -0.090548
2020   -0.035157
2030   -0.067454
2510   -0.207992
2520   -0.208598
2530   -0.062583
2550   -0.151223
3010    0.078312
3020    0.038780
3030   -0.153771
3510   -0.066432
3520   -0.163937
4010   -0.087832
4020   -0.122515
4030    0.003779
4510   -0.108644
4520   -0.172065
4530   -0.252813
5010    0.011838
5020   -0.070871
5510    0.040759
6010   -0.035826
Name: RetYTD, dtype: float64

### Create industrial indicators

In [ ]:
df = pd.get_dummies(df, columns=['ggroup'])
df

### Do the highest and lowest return industries make economic sense for early 2022 (RetYTD)? Explain in detail.

In [ ]:
# get min value
avg.min()

-0.2528127272727273

In [ ]:
# # get min value index
avg.idxmin()

4530

In [ ]:
# # get max value
avg.max()

0.5150346666666666

In [ ]:
# # get max value index
avg.idxmax()

1010

We can find the highest return industry is 1010(Energy)(average return is 0.51), and the lowest return industry is 4530(Semiconductors & Semiconductor Equipment)(average return is -0.25).

We can conclude that according to the return data of early 2022, investing in Energy industry would have a great profit. On the contrary, investing in Semiconductor industry would have loss. 

### Run fixed effect regressions (using industry fixed effects) to explain companies' stock returns in early 2022 (RetYTD):

In [ ]:
# add constant column to the original dataframe
df['constant'] = 1

In [ ]:
df.columns

Index(['gvkey', 'datadate', 'fyear', 'indfmt', 'consol', 'popsrc', 'datafmt',
       'Ticker', 'curcd', 'costat', 'Name ', 'RetYTD', 'ggroup_1010',
       'ggroup_1510', 'ggroup_2010', 'ggroup_2020', 'ggroup_2030',
       'ggroup_2510', 'ggroup_2520', 'ggroup_2530', 'ggroup_2550',
       'ggroup_3010', 'ggroup_3020', 'ggroup_3030', 'ggroup_3510',
       'ggroup_3520', 'ggroup_4010', 'ggroup_4020', 'ggroup_4030',
       'ggroup_4510', 'ggroup_4520', 'ggroup_4530', 'ggroup_5010',
       'ggroup_5020', 'ggroup_5510', 'ggroup_6010'],
      dtype='object')

In [ ]:
# define x as a subset of original dataframe
x = df.drop(columns=['gvkey', 'datadate','fyear','indfmt','consol','popsrc','datafmt','Ticker','curcd','costat','Name ','RetYTD','constant'])
# define y as a series
y = df['RetYTD']

# pass x as a dataframe, while pass y as a series
model = sm.OLS(y, x).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 RetYTD   R-squared:                       0.326
Model:                            OLS   Adj. R-squared:                  0.317
Method:                 Least Squares   F-statistic:                     39.07
Date:                Mon, 25 Apr 2022   Prob (F-statistic):          6.29e-141
Time:                        13:58:32   Log-Likelihood:                 357.63
No. Observations:                1886   AIC:                            -667.3
Df Residuals:                    1862   BIC:                            -534.2
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
ggroup_1010     0.5150      0.023     22.140      0.000       0.469       0.561
ggroup_1510     0.0578      0.021      2.799      0.005       0.017       0.098
ggroup_2010    -0.0905      0.015     -6.080      0.000      -0.120      -0.061
ggroup_2020    -0.0352      0.025     -1.385      0.166      -0.085       0.015
ggroup_2030    -0.0675      0.032     -2.091      0.037      -0.131      -0.004
ggroup_2510    -0.2080      0.040     -5.162      0.000      -0.287      -0.129
ggroup_2520    -0.2086      0.025     -8.218      0.000      -0.258      -0.159
ggroup_2530    -0.0626      0.025     -2.485      0.013      -0.112      -0.013
ggroup_2550    -0.1512      0.022     -6.961      0.000      -0.194      -0.109
ggroup_3010     0.0783      0.049      1.603      0.109      -0.018       0.174
ggroup_3020     0.0388      0.030      1.277      0.202      -0.021       0.098
ggroup_3030    -0.1538      0.049     -3.147      0.002      -0.250      -0.058
ggroup_3510    -0.0664      0.018     -3.627      0.000      -0.102      -0.031
ggroup_3520    -0.1639      0.016    -10.389      0.000      -0.195      -0.133
ggroup_4010    -0.0878      0.015     -5.833      0.000      -0.117      -0.058
ggroup_4020    -0.1225      0.021     -5.833      0.000      -0.164      -0.081
ggroup_4030     0.0038      0.026      0.146      0.884      -0.047       0.054
ggroup_4510    -0.1086      0.019     -5.858      0.000      -0.145      -0.072
ggroup_4520    -0.1721      0.022     -7.734      0.000      -0.216      -0.128
ggroup_4530    -0.2528      0.027     -9.307      0.000      -0.306      -0.200
ggroup_5010     0.0118      0.056      0.212      0.832      -0.098       0.121
ggroup_5020    -0.0709      0.031     -2.280      0.023      -0.132      -0.010
ggroup_5510     0.0408      0.026      1.541      0.124      -0.011       0.093
ggroup_6010    -0.0358      0.018     -2.035      0.042      -0.070      -0.001
==============================================================================
Omnibus:                      403.265   Durbin-Watson:                   2.008
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2314.656
Skew:                           0.877   Prob(JB):                         0.00
Kurtosis:                       8.136   Cond. No.                         3.75
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

According to the regression, we can find the coefficient of 
*   ggroup_1010 is 0.5150
*   ggroup_1510 is 0.0578
*   ggroup_2010 is -0.0905
*   ggroup_2020 is -0.0352
*   ggroup_2030 is -0.0675
*   ggroup_2510 is -0.2080
*   ggroup_2520 is -0.2086
*   ggroup_2530 is -0.0626
*   ggroup_2550 is -0.1512
*   ggroup_3010 is 0.0783
*   ggroup_3020 is 0.0388
*   ggroup_3030 is -0.1538
*   ggroup_3510 is -0.0664
*   ggroup_3520 is -0.1639
*   ggroup_4010 is -0.0878
*   ggroup_4020 is -0.1225
*   ggroup_4030 is 0.0038
*   ggroup_4510 is -0.1086
*   ggroup_4520 is -0.1721
*   ggroup_4530 is -0.2528
*   ggroup_5010 is 0.0118
*   ggroup_5020 is -0.0709
*   ggroup_5510 is 0.0408
*   ggroup_6010 is -0.0358
Also, R-squared is 0.326 and Adj. R-squared is 0.317. 
We can conclude that the whole regression's explanatory is not good because the R-squared and Adj. R-squared are below 0.5. 

Many ggroups' coefficient is negative means the they have negative effect on the target variable (Stock return). Besides, there are six P-values bigger than 0.05, which means those ggroups' results are not statistical significant.